compiler/il/symbol/OMRResolvedMethodSymbol.cpp

```diff
@@ -1998,8 +1998,24 @@ OMR::ResolvedMethodSymbol::detectInternalCycles(TR::CFG *cfg, TR::Compilation *c
                            }
                         }
 
-                     //rip out trees
-                     clonedCatch->getEntry()->join(clonedCatch->getExit());
+                     TR::TreeTop *retain = clonedCatch->getEntry();
+                     // As this method is performed soon after ilgen, the exception handler
+                     // may be prepended with an asynccheck and pending pushes
+                     // These should be retained in the copy, so skip them when ripping out trees
+                     if (comp->getHCRMode() == TR::osr)
+                        {
+                        TR::TreeTop *next = retain->getNextTreeTop();
+                        while (next &&
+                               (next->getNode()->getOpCodeValue() == TR::asynccheck ||
+                               (next->getNode()->getSymbolReference() &&
```

Comment:
We should probably call hasSymbolReference() on the opcode first before getting the symbol reference (to be safe). 
Might also be safer to more explicitly check for an auto store rather than the somewhat more generic CP index < 0 test here.